File: src/main/java/org/oewntk/yaml/in/LexParser.java
/*
 * Copyright (c) $originalComment.match("Copyright \(c\) (\d+)", 1, "-")2021. Bernard Bou.
 */

package org.oewntk.yaml.in;

import org.oewntk.model.Lex;
import org.oewntk.model.Pronunciation;
import org.oewntk.model.Sense;

import java.io.File;
import java.util.*;
import java.util.Map.Entry;

/**
 * Lex parser
 */
public class LexParser extends YamProcessor<Lex, String, Map<String, Object>>
{
    private static final boolean DUMP = false;

    private static final String KEY_LEX_SENSE = "sense";
    private static final String KEY_LEX_PRONUNCIATION = "pronunciation";
    private static final String KEY_LEX_FORM = "form";

    private static final String KEY_SENSE_ID = "id";
    private static final String KEY_SENSE_SYNSET = "synset";
    private static final String KEY_SENSE_ADJPOSITION = "adjposition";
    private static final String KEY_SENSE_VERBFRAMES = "subcat";
    private static final String KEY_SENSE_EXAMPLES = "sent";

    private static final String KEY_PRONUNCIATION_VARIETY = "variety";
    private static final String KEY_PRONUNCIATION_VALUE = "value";

    private static final String[] VALID_SENSE_RELATIONS = { //
            "antonym", //
            "similar",  //
            "exemplifies",  //
            "derivation",  //
            "pertainym",  //
            "participle",  //
            "also",  //
            "domain_region",  //
            "domain_topic",  //

            "state",
            "result",
            "event",
            "property",
            "location",
            "destination",
            "agent",
            "undergoer",
            "uses",
            "instrument",
             "by_means_of",
            "material",
            "vehicle",
            "body_part",

            "other"};

    private static final String[] SENSE_RELATIONS = { //
            "antonym", //
            "similar",  //
            "exemplifies", "is_exemplified_by", //
            "derivation",  //
            "pertainym",  //
            "participle",  //
            "also",  //
            "domain_region", "has_domain_region", //
            "domain_topic", "has_domain_topic", //

            "agent",
            "material",
            "event",
            "instrument",
            "location",
            "by_means_of",
            "undergoer",
            "property",
            "result",
            "state",
            "uses",
            "destination",
            "body_part",
            "vehicle",

            "other"
    };

    private static final String[] VOID_STRING_ARRAY = new String[0];

    /**
     * Accumulated senses as lexes are processed
     */
    private final List<Sense> senses = new ArrayList<>();

    /**
     * Lex YAML parser
     *
     * @param dir dir containing YAML files
     */
    public LexParser(final File dir)
    {
        super(dir);
    }

    @Override
    protected File[] getFiles()
    {
        return dir.listFiles((f) -> f.getName().matches("entries.*\\.yaml"));
    }

    @Override
    protected Collection<Lex> processEntry(final String source, final Entry<String, Map<String, Object>> lemmaEntry)
    {
        List<Lex> lexes = new ArrayList<>();

        String lemma = lemmaEntry.getKey();
        Map<String, Object> lemmaMap = lemmaEntry.getValue();

        for (Map.Entry<String, Object> typeEntry : lemmaMap.entrySet())
        {
            String type = typeEntry.getKey();
            Map<String, Object> lexMap = (Map<String, Object>) typeEntry.getValue();
            YamlUtils.assertKeysIn(source, lexMap.keySet(), KEY_LEX_SENSE, KEY_LEX_PRONUNCIATION, KEY_LEX_FORM);
            if (DUMP)
            {
                YamlUtils.dumpMap("%s %s%n", lexMap);
            }

            // lex
            Lex lex = new Lex(lemma, type, source);

            // pronunciations
            List<Map<String, Object>> pronunciationList = (List<Map<String, Object>>) lexMap.get(KEY_LEX_PRONUNCIATION);
            Pronunciation[] pronunciations = null;
            if (pronunciationList != null)
            {
                pronunciations = new Pronunciation[pronunciationList.size()];
                int p = 0;
                for (Map<String, Object> pronunciationMap : pronunciationList)
                {
                    YamlUtils.assertKeysIn(source, pronunciationMap.keySet(), KEY_PRONUNCIATION_VARIETY, KEY_PRONUNCIATION_VALUE);
                    if (DUMP)
                    {
                        YamlUtils.dumpMap("\t%s %s%n", pronunciationMap);
                    }
                    String variety = (String) pronunciationMap.get(KEY_PRONUNCIATION_VARIETY);
                    String value = (String) pronunciationMap.get(KEY_PRONUNCIATION_VALUE);
                    pronunciations[p] = new Pronunciation(value, variety);
                    p++;
                }
            }
            lex.setPronunciations(pronunciations);

            // forms
            List<String> forms = (List<String>) lexMap.get(KEY_LEX_FORM);
            if (forms != null)
            {
                lex.setForms(forms.toArray(new String[0]));
            }

            // senses
            List<Map<String, Object>> senseMaps = (List<Map<String, Object>>) lexMap.get(KEY_LEX_SENSE);
            Sense[] lexSenses = new Sense[senseMaps.size()];
            int i = 0;
            for (Map<String, Object> senseMap : senseMaps)
            {
                YamlUtils.assertKeysIn(source, senseMap.keySet(),
                        VALID_SENSE_RELATIONS, //
                        KEY_SENSE_ID, //
                        KEY_SENSE_SYNSET, //
                        KEY_SENSE_VERBFRAMES, //
                        KEY_SENSE_VERBFRAMES, //
                        KEY_SENSE_ADJPOSITION,
                        KEY_SENSE_EXAMPLES //
                );
                if (DUMP)
                {
                    YamlUtils.dumpMap("\t%s %s%n", senseMap);
                }

                String senseId = (String) senseMap.get(KEY_SENSE_ID);
                String synsetId = (String) senseMap.get(KEY_SENSE_SYNSET);
                List<String> examplesList = (List<String>) senseMap.get(KEY_SENSE_EXAMPLES);
                List<String> verbFramesList = (List<String>) senseMap.get(KEY_SENSE_VERBFRAMES);
                String[] examples = examplesList == null ? null : examplesList.toArray(VOID_STRING_ARRAY);
                String[] verbFrames = verbFramesList == null ? null : verbFramesList.toArray(VOID_STRING_ARRAY);
                String adjPosition = (String) senseMap.get(KEY_SENSE_ADJPOSITION);

                // relations
                Map<String, Set<String>> relations = null;
                for (String relationKey : SENSE_RELATIONS)
                {
                    if (senseMap.containsKey(relationKey))
                    {
                        List<String> relationTargets = (List<String>) senseMap.get(relationKey);
                        if (relations == null)
                        {
                            relations = new TreeMap<>();
                        }
                        relations.computeIfAbsent(relationKey, (k) -> new LinkedHashSet<>()).addAll(relationTargets);
                    }
                }

                // sense
                lexSenses[i] = new Sense(senseId, lex, type.charAt(0), i, synsetId, examples, verbFrames, adjPosition, relations);
                senses.add(lexSenses[i]);

                i++;
            }
            lex.setSenses(lexSenses);

            // accumulate
            lexes.add(lex);
        }

        return lexes;
    }

    /**
     * Get senses that was also accumulated
     *
     * @return accumulated senses
     */
    public Collection<Sense> getSenses()
    {
        return this.senses;
    }
}

File: src/main/java/org/oewntk/yaml/in/LexParser.kt
/*
 * Copyright (c) 2024. Bernard Bou.
 */
package org.oewntk.yaml.`in`

import org.oewntk.model.*
import org.oewntk.model.Sense.Companion.SENSE_RELATIONS
import org.oewntk.model.Sense.Companion.VALID_SENSE_RELATIONS
import org.oewntk.yaml.`in`.YamlUtils.assertKeysIn
import org.oewntk.yaml.`in`.YamlUtils.processExamples
import org.oewntk.yaml.`in`.YamlUtils.safeCast
import org.oewntk.yaml.`in`.YamlUtils.safeNullableCast
import java.io.File

/**
 * Lex YAML parser
 *
 * @property dir dir containing YAML files
 */
class LexParser(
    dir: File,
    val fileext: String = "yaml",
    verbose: Boolean = false,
    val throws: Boolean = true,
) : YamlProcessor<Lex, String, Map<String, *>>(dir, verbose = verbose) {

    /**
     * Accumulated senses as lexes are processed
     */
    val senses: MutableList<Sense> = ArrayList()

    override val files: Array<File>
        get() = dir.listFiles { f: File -> f.name.matches("entries.*\\.$fileext".toRegex()) }!!

    override fun processEntry(source: String?, entry: Pair<String, Map<String, *>>): Collection<Lex> {
        try {
            val lexes: MutableList<Lex> = ArrayList()

            val lemma = entry.first
            val lemmaMap = entry.second
            for ((k2, value1) in lemmaMap) {
                val type = SynsetType.fromChar(k2[0])
                val lexMap: Map<String, *> = safeCast(value1!!)
                assertKeysIn(
                    throws = throws,
                    lexMap.keys,
                    KEY_LEX_SENSE,
                    KEY_LEX_PRONUNCIATION,
                    KEY_LEX_FORM,
                )
                if (DUMP) {
                    dumpMap(lexMap)
                }

                // lex
                val lex = Lex(lemma, k2)

                // senses
                val senseMaps: List<Map<String, *>> = safeCast(lexMap[KEY_LEX_SENSE]!!)
                val lexSenses = MutableList(senseMaps.size) {
                    val senseMap = senseMaps[it]
                    assertKeysIn(
                        throws = throws,
                        senseMap.keys,
                        VALID_SENSE_RELATIONS,
                        KEY_SENSE_ID,
                        KEY_SENSE_SYNSET,
                        KEY_SENSE_VERBFRAMES,
                        KEY_SENSE_VERBTEMPLATES,
                        KEY_SENSE_ADJPOSITION,
                        KEY_SENSE_TAGCOUNT,
                        KEY_SENSE_EXAMPLES,
                    )
                    if (DUMP) {
                        dumpMap(senseMap, indent = "\t")
                    }

                    val senseId = senseMap[KEY_SENSE_ID]!! as SenseKey
                    val synsetId = senseMap[KEY_SENSE_SYNSET]!! as SynsetId
                    val examples: List<Pair<String, String?>>? = processExamples(
                        safeNullableCast(senseMap[KEY_SENSE_EXAMPLES]),
                        KEY_EXAMPLE_TEXT,
                        KEY_EXAMPLE_SOURCE
                    )
                    val verbFrames: List<VerbFrameId>? = safeNullableCast(senseMap[KEY_SENSE_VERBFRAMES])
                    val verbTemplates: List<VerbTemplateId>? = safeNullableCast(senseMap[KEY_SENSE_VERBTEMPLATES])
                    val adjPosition = senseMap[KEY_SENSE_ADJPOSITION] as String?
                    val tagCount = senseMap[KEY_SENSE_TAGCOUNT] as Int?

                    // relations
                    val relations = if (IGNORE_QTARGETS)
                        SENSE_RELATIONS
                            .asSequence()
                            .filter { relation -> senseMap.containsKey(relation) }
                            .associateWith { relation ->
                                safeCast<List<String>>(senseMap[relation]!!).toSet()
                            }
                            .ifEmpty { null }
                    else
                        SENSE_RELATIONS
                            .asSequence()
                            .filter { relation -> senseMap.containsKey(relation) }
                            .map { relation ->
                                relation to safeCast<List<String>>(senseMap[relation]!!).filter { target -> target[0] != 'Q' }
                                    .toSet()
                            }
                            .filter { (_, targets) -> targets.isNotEmpty() }
                            .toMap()
                            .ifEmpty { null }

                    // sense
                    val lexSense = Sense(
                        senseId,
                        lex.key,
                        synsetId,
                        indexInLex = it,
                        examples?.toTypedArray(),
                        verbFrames?.toTypedArray(),
                        verbTemplates?.toTypedArray(),
                        adjPosition,
                        tagCount,
                        relations
                    )
                    senses.add(lexSense)
                    lexSense
                }
                lex.senseKeys = lex.senseKeys.toMutableList() + lexSenses.map { it.senseKey }.asSequence()

                // pronunciations
                val pronunciationList: List<Map<String, *>>? = safeNullableCast(lexMap[KEY_LEX_PRONUNCIATION])
                val pronunciations: Array<Pronunciation>? = if (pronunciationList != null) {
                    Array(pronunciationList.size) {
                        val pronunciationMap = pronunciationList[it]
                        assertKeysIn(
                            throws = throws,
                            pronunciationMap.keys,
                            KEY_PRONUNCIATION_VARIETY,
                            KEY_PRONUNCIATION_VALUE
                        )
                        if (DUMP) {
                            dumpMap(pronunciationMap, indent = "\t")
                        }
                        val variety = pronunciationMap[KEY_PRONUNCIATION_VARIETY] as String?
                        val value = pronunciationMap[KEY_PRONUNCIATION_VALUE] as String
                        Pronunciation(value, variety)
                    }
                } else null
                lex.pronunciations = pronunciations?.toSet()

                // forms
                val forms: List<String>? = safeNullableCast(lexMap[KEY_LEX_FORM])
                lex.forms = forms?.toSet() // preserves order (LinkedHashSet)

                // accumulate
                lexes.add(lex)
            }
            return lexes

        } catch (iae: IllegalArgumentException) {
            throw IllegalArgumentException("${iae.message} in $source")
        }
    }

    companion object {

        private const val IGNORE_QTARGETS = true

        private const val DUMP = false

        private const val KEY_LEX_SENSE = "sense"
        private const val KEY_LEX_PRONUNCIATION = "pronunciation"
        private const val KEY_LEX_FORM = "form"

        private const val KEY_SENSE_ID = "id"
        private const val KEY_SENSE_SYNSET = "synset"
        private const val KEY_SENSE_ADJPOSITION = "adjposition"
        private const val KEY_SENSE_VERBFRAMES = "subcat"
        private const val KEY_SENSE_EXAMPLES = "sent"
        private const val KEY_SENSE_VERBTEMPLATES = "template" // not compat
        private const val KEY_SENSE_TAGCOUNT = "tagcount" // not compat

        private const val KEY_EXAMPLE_SOURCE = "source"
        private const val KEY_EXAMPLE_TEXT = "text"

        private const val KEY_PRONUNCIATION_VARIETY = "variety"
        private const val KEY_PRONUNCIATION_VALUE = "value"
    }
}
